lib/controllers/messages.ts

```diff
@@ -343,15 +357,25 @@ export function CreateMessageHandler(
       return ResponseErrorForbiddenNotAuthorizedForDefaultAddresses;
     }
 
+    const timeToLive = TimeToLive.decode(messagePayload.time_to_live);
```

Comment:
That's because messagePayload carries an "ApiNewMessage" type where time_to_live is optional. As it's not guaranted that the middleware gets called before this point (ie. in tests), there is a double check here.

Reply:
moreover typescript want that we assert that ttl is not undefined, so if you omit the decoding here you have to put one more if

Reply:
mmh no wait, the function signature guarantees to have a valid `ApiNewMessage` (aka `NewMessage`) as input (whether it comes from `MessagePayloadMiddleware` or something else is irrelevant) - and [NewMessage](https://github.com/teamdigitale/digital-citizenship-functions/blob/ae38da76c73362c082e799ca6d78b3d2ed3bad19/lib/api/definitions/NewMessage.ts#L27) is guaranteed to have `time_to_live` of type `TimeToLive | undefined` - so the right thing here is to just check whether `time_to_live` is defined or not, no need to validate, that *must* already have happened before.

Reply:
indeed you have to check here if ttl is undefined AND, if it is, use the default value (which is what decode() does)

Reply:
ahhh that's the problem then, this function doesn't want a `NewMessage`, this function wants a `NewMessage` where `undefined` values have been replaced by their defaults

the "problem" here is that `NewMessage` has `time_to_live` as optional when in fact is a required field (set to a default when not provided by the client)

it shouldn't be the concern of this function to decode payloads and replace undefined values with defaults, that must happen *before* by the middleware... and the input type should guarantee what this function expects

Reply:
Just keep this in mind: decoding and validation must happen only at the system boundary (e.g. with network or disk) - the responsibility of a middleware is to do this validation (decode `any` to a well defined type), ensuring that their output doesn't ever need to be validated again (indeed this is possible only when a type system can carry this information, like in Typescript).

So when you find yourself to have to do a validation inside the system, that is a smell of incorrect design of types.

Reply:
Just keep this in mind: decoding and validation must happen only at the system boundary (e.g. with network or disk) - the responsibility of a middleware is to do this validation (decode `any` to a well defined type), ensuring that their output doesn't ever need to be validated again (indeed this is possible only when a type system can carry this information, like in Typescript).

So when you find yourself to have to do a validation inside the system, that is a smell of incorrect design of types.

Reply:
Just keep this in mind: decoding and validation must happen only at the system boundary (e.g. with network or disk) - the responsibility of a middleware is to do this validation (decode `any` to a well defined type), ensuring that their output doesn't ever need to be validated again (indeed this is possible only when a type system can carry this information, like in Typescript).

So when you find yourself to have to do a validation inside the system, that is a smell of incorrect design of types.

Reply:
refactored, see https://github.com/teamdigitale/digital-citizenship-functions/pull/158/files#diff-ca9342512f1db78ecb71b1a86cd82b77R113

Anyway the interface is rather clumsy because we have to keep this duality between input types (ttl optional, comes from the API call) and output types (ttl mandatory, to feed CreateMessageHandler) even if we know that calling ApiMessage.decode() will populate the ttl field (that's why we need another type ApiMessageWithDefaults)